ompi/mca/pml/ucx/pml_ucx.c

```diff
@@ -749,6 +744,74 @@ int mca_pml_ucx_send(const void *buf, size_t count, ompi_datatype_t *datatype, i
     }
 }
 
+#if (HAVE_DECL_UCP_TAG_SEND_NBR == 1)
+static __opal_attribute_noinline__ int
+mca_pml_ucx_send_nb_noinline(ucp_ep_h ep, const void *buf, size_t count,
+                             ompi_datatype_t *datatype, ucp_datatype_t ucx_datatype,
+                             ucp_tag_t tag, mca_pml_base_send_mode_t mode,
+                             ucp_send_callback_t cb)
+{
+    return mca_pml_ucx_send_nb(ep, buf, count, datatype, ucx_datatype, tag, mode, cb);
+}
+
+static inline __opal_attribute_always_inline__ int
+mca_pml_ucx_send_nbr(ucp_ep_h ep, const void *buf, size_t count,
+                     ucp_datatype_t ucx_datatype, ucp_tag_t tag)
+
+{
+    void *req;
+    ucs_status_t status;
+
+    req    = PML_UCX_REQ_ALLOCA();
+    status = ucp_tag_send_nbr(ep, buf, count, ucx_datatype, tag, req);
+    if (OPAL_LIKELY(status == UCS_OK)) {
+        return OMPI_SUCCESS;
+    }
+
+    ucp_worker_progress(ompi_pml_ucx.ucp_worker);
+    while ((status = ucp_request_check_status(req)) == UCS_INPROGRESS) {
+        opal_progress();
+    }
+
+    return OPAL_LIKELY(UCS_OK == status) ? OMPI_SUCCESS : OMPI_ERROR;
+}
+#endif
+
+int mca_pml_ucx_send(const void *buf, size_t count, ompi_datatype_t *datatype, int dst,
+                     int tag, mca_pml_base_send_mode_t mode,
+                     struct ompi_communicator_t* comm)
+{
+    ucp_ep_h ep;
+
+    PML_UCX_TRACE_SEND("%s", buf, count, datatype, dst, tag, mode, comm,
+                       mode == MCA_PML_BASE_SEND_BUFFERED ? "bsend" : "send");
+
+    ep = mca_pml_ucx_get_ep(comm, dst);
+    if (OPAL_UNLIKELY(NULL == ep)) {
+        PML_UCX_ERROR("Failed to get ep for rank %d", dst);
+        return OMPI_ERROR;
+    }
+
+#if (HAVE_DECL_UCP_TAG_SEND_NBR == 1)
+    if (OPAL_LIKELY(MCA_PML_BASE_SEND_BUFFERED != mode &&
+                    MCA_PML_BASE_SEND_SYNCHRONOUS != mode)) {
+        return mca_pml_ucx_send_nbr(ep, buf, count,
+                                    mca_pml_ucx_get_datatype(datatype),
+                                    PML_UCX_MAKE_SEND_TAG(tag, comm));
+    }
+
+    return mca_pml_ucx_send_nb_noinline(ep, buf, count, datatype,
+                                        mca_pml_ucx_get_datatype(datatype),
+                                        PML_UCX_MAKE_SEND_TAG(tag, comm), mode,
+                                        mca_pml_ucx_send_completion);
+#endif
+
+    return mca_pml_ucx_send_nb(ep, buf, count, datatype,
```

Comment:
how about making mca_pml_ucx_send_nb inline even in case mca_pml_ucx_send_nbr() is present, and put its code here?
it will reduce code complexity
can you pls check how it affects performance?